perllib/FixMyStreet/Roles/Extra.pm

```diff
@@ -0,0 +1,170 @@
+
+And
+
+    # e.g. for sites like Zurich (but handled gracefully otherwise)
+    $problem->set_extra_metadata($c, overdue => 1 );
+    if ($problem->get_extra_metadata($c, 'overdue')) { ... }
+
+=head1 METHODS
+
+=head2 set_extra_metadata
+
+    $problem->set_extra_metadata($c, overdue => 1);
+
+=cut
+
+sub set_extra_metadata {
+    my ($self, $c, $key, $value) = @_;
+    my $extra = $self->get_extra($c);
+
+    return unless ref $extra eq 'HASH';
+    $self->extra({ %$extra, $key => $value });
+};
+
+=head2 unset_extra_metadata
+
+    $contact->unset_extra_metadata($c, 'photo_required');
+
+=cut
+
+sub unset_extra_metadata {
+    my ($self, $c, $key) = @_;
+    my $extra = $self->get_extra($c);
+
+    return unless ref $extra eq 'HASH';
+    return 1 unless exists $extra->{$key};
+    delete $extra->{$key};
+    $self->extra($extra);
+};
+
+=head2 set_extra_metadata
+
+    my $overdue = $problem->get_extra_metadata($c, 'overdue');
+
+=cut
+
+sub get_extra_metadata {
+    my ($self, $c, $key) = @_;
+    my $extra = $self->get_extra($c);
+
+    return unless ref $extra eq 'HASH';
+    return $extra->{$key};
+};
+
+=head2 get_extra_metadata_as_hashref
+
+    my $hashref = $contact->get_extra_metadata_as_hashref($c);
+
+=cut
+
+my $META_FIELD = '_fields';
+
+sub get_extra_metadata_as_hashref {
+    my ($self, $c) = @_;
+    my $extra = $self->get_extra($c);
+
+    return {} unless ref $extra eq 'HASH';
+    my %extra = %$extra;
+    delete $extra{$META_FIELD};
+    return \%extra;
+}
+
+=head2 get_extra_fields
+
+    my $metas = $problem->get_extra_fields($c);
+
+=cut
+
+sub get_extra_fields {
+    my ($self, $c) = @_;
+    my $extra = $self->get_extra($c);
+
+    return $extra if ref $extra eq 'ARRAY';
+    return $extra->{$META_FIELD} ||= do {
+        $self->dirty_extra;
+        # not sure this will work transparently with RABX encoding
+        [];
+    };
+}
+
+=head2 set_extra_fields
+
+    $problem->set_extra_fields($c, { ... }, { ... } );
+
+=cut
+
+sub set_extra_fields {
+    my ($self, $c, @fields) = @_;
+    my $extra = $self->get_extra($c);
+
+    if (ref $extra eq 'ARRAY') {
+        # replace extra entirely with provided fields
+        $self->extra(\@fields);
+    }
+    else {
+        $self->extra({ %$extra, $META_FIELD => \@fields });
+    }
+}
+
+=head1 HELPER METHODS
+
+For internal use mostly.
+
+=head2 dirty_extra
+
+Set the extra field as dirty.  (e.g. signalling that the DB object should be
+updated).
+
+=cut
+
+sub dirty_extra {
+    my $self = shift;
+    $self->make_column_dirty('extra');
+    return 1;
+}
+
+=head2 get_extra
+
+Get the extra data.  If this is not set, then returns a {} or [] reference
+I<as appropriate>, e.g. depending on the Cobrand's expected layout.  This
+worrying about whether we should return a hash or array reference is the reason
+that we have to pass $c to Every Single Method (i.e. so that we can get access
+to C<$c-E<gt>cobrand>.
+
+(This is a bit of a wart, can we do better?)
```

Comment:
As this doesn't need to worry about problem cobrand not matching site cobrand, you could perhaps access the cobrand on $self->cobrand (though it won't be a class). However, I wonder if it would be worth changing the other places in the codebase that assume extra to be an arrayref to use new functions and have get_extra always return a hashref. It would simplify much of the code of this file, and there only appears to be four or five places to do it. In fact, could you then also use an `after` on Problem's `extra` rather than a separate `get_extra` function, so that it always returned a hashref regardless of what old data might be in the database.